### PR TITLE
Checkbox: Adding base prototype. 

### PR DESCRIPTION
**Pull request by aneeshack4**
Made this prototype for the `Checkbox` component to test out some of the theories from the [spec](https://github.com/microsoft/fluent-ui-react-staging/tree/fluent_checkbox/packages/base/src/components/Checkbox) and sending this out to get feedback and iterate.

### CODE REVIEW

File: packages/base/src/components/Checkbox/Checkbox.base.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+export const CheckboxBase: React.FunctionComponent<ICheckboxProps> = (
+  props: ICheckboxProps
+) => {
+  const { children, slots } = props;
+    const {
+        root: Root = "label",
+        input: Input = "input",
+        icon: Icon = "i" 
+
```
Comment (khmakoto) [inline packages/base/src/components/Checkbox/Checkbox.base.tsx]: Why do we have this extra line break here?

File: packages/base/src/components/Checkbox/Checkbox.types.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  /** Defines the children of the Checkbox component. */
+  children?: React.ReactNode;
+
+  /** Defines whether the Checkbox is in an enabled or disabled state. */
+  disabled?: boolean;
+  
+  /** Defines whether default value of the checkbox is checked or unchecked. (Controlled) */
+  checked?: boolean;
+
+  /** default value (uncontrolled) */
```
Comment (khmakoto) [inline packages/base/src/components/Checkbox/Checkbox.types.ts]:
```suggestion
  /** Defines the default value of the checkbox for uncontrolled scenarios. */
```

File: packages/base/src/components/Checkbox/Checkbox.types.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  defaultChecked?: boolean;
+  
+  /** Defines a callback that is triggered when the Checkbox is toggled. */
+  onChange?: (ev: MouseEvent, checked: boolean) => void;
+}
```
Comment (khmakoto) [inline packages/base/src/components/Checkbox/Checkbox.types.ts]: I haven't really talked about this but I feel that it'd be good to have our interface props be defined in alphabetical order. Maybe this is as good a place as any to start?

File: packages/base/src/components/Checkbox/useCheckbox.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import { ICheckboxProps } from "./Checkbox.types";
+
+export interface ICheckboxState {
+  onClick: (ev: MouseEvent, checked: boolean) => void;
+  isChecked: boolean;
+  rootRef: React.Ref<Element>;
+}
+
+const useCheckboxState = (userProps: ICheckboxProps): ICheckboxState => {
+  const { disabled, onChange, checked: controlledValue, defaultChecked } = userProps;
```
Comment (khmakoto) [inline packages/base/src/components/Checkbox/useCheckbox.ts]: Maybe also destructure in alphabetical order?

File: packages/base/src/components/Checkbox/useCheckbox.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+  const [isChecked, setChecked] = useControlledState(controlledValue, defaultChecked);
+
+  const rootRef = React.useRef<HTMLElement>(null);
+
+  const onCheckboxChange = (ev: MouseEvent, checked: boolean) => {
+    if (!disabled && onChange) {
+      onChange(ev, !checked);
+      
+      setChecked(!isChecked);
```
Comment (khmakoto) [inline packages/base/src/components/Checkbox/useCheckbox.ts]: Should you really be using `isChecked` here?

**Comment by khmakoto** [comment]
Also, you need to run `yarn change` to pass the build now.

File: change/@fluentui-base-2019-10-31-15-38-12-checkbox_proto.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{
+  "type": "minor",
+  "comment": "Checkbox proto",
```
Comment (khmakoto) [inline change/@fluentui-base-2019-10-31-15-38-12-checkbox_proto.json]:
```suggestion
  "comment": "Checkbox: Adding base prototype.",
```

File: packages/base/src/components/Checkbox/useCheckbox.ts (hunk 5 above)
Comment (aneeshack4) [reply packages/base/src/components/Checkbox/useCheckbox.ts]: It's the `state` variable so I'm setting it on change - I'm using `checked` from props as the `controlledValue`.

Comment (aneeshack4) [reply packages/base/src/components/Checkbox/useCheckbox.ts]: Maybe I should be passing in `isChecked` to the onChange call right before this tho...

**Comment by aneeshack4** [comment]
the failure is due to a linter error on the console statement in the stories file... Why are we not allowing this?

File: packages/base/src/components/Checkbox/useCheckbox.ts (hunk 5 above)
Comment (khmakoto) [reply packages/base/src/components/Checkbox/useCheckbox.ts]: Yeah, I was just confused as to why they were different.

**Comment by khmakoto** [comment]
Ahh yeah, I had that error as well, I disabled that for the next line in `Button.stories.tsx` but this might be a good moment to disable this lint rule. What do you think @kenotron?

File: packages/base/src/components/Checkbox/useCheckbox.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+const useCheckboxState = (userProps: ICheckboxProps): ICheckboxState => {
+  const { checked: controlledValue, defaultChecked, disabled, onChange } = userProps;
+
+  const [isChecked, setChecked] = useControlledState(controlledValue, defaultChecked);
+
+  const rootRef = React.useRef<HTMLElement>(null);
+
+  const onCheckboxChange = (ev: MouseEvent, checked: boolean) => {
+    if (!disabled && onChange) {
+      onChange(ev, !checked);
```
Comment (khmakoto) [inline packages/base/src/components/Checkbox/useCheckbox.ts]: Just change this to `isChecked` as we mentioned in the other thread and I'll accept 😄

Comment (aneeshack4) [reply packages/base/src/components/Checkbox/useCheckbox.ts]: done! sorry missed that

**Pull request merged**